tests/sota_tools/treehub_server.py

```diff
@@ -1,85 +1,155 @@
-            return True
+        self.__class__.made_requests += 1
+        if args.fail and args.fail > 0:
+            return self.__class__.made_requests % args.fail == 0
         else:
-            if drop_connection_every_n_request == self.made_requests[self.path]:
-                self.made_requests[self.path] = 0
             return False
 
 
-class ReUseHTTPServer(HTTPServer):
-    def server_bind(self):
-        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        HTTPServer.server_bind(self)
+def sig_handler(signum, frame):
+    sys.exit(0)
 
 
-server_address = ('', int(sys.argv[1]))
-httpd = ReUseHTTPServer(server_address, TreehubServerHandler)
-repo_path = "tests/sota_tools/repo"
-if len(sys.argv) == 3 and sys.argv[2]:
-    print("Dropping connection after every: %s request" % sys.argv[2])
-    drop_connection_every_n_request = int(sys.argv[2])
-else:
-    drop_connection_every_n_request = 0
+def create_repo(path):
+    """
+    Creates a new ostree repository with persistent object checksums.
+    To achive persistency, we generate files with the same seed(0).
+    As OSTree content objects objects include uid, gid, and extended attributes,
+    we strip the extended attributes and set the rest to constant values.
+    Same should be done for permissions, stored in dirmeta objects, and
+    timestamp in commit.
+    """
+    tree = Path(path) / 'tree'
+    tree.mkdir(mode=0o755)
+    seed(0)  # to generate same files each time
+    try:
+        for i in range(10):
+            file = (tree / str(i)).with_suffix('.bin')
+            file.write_bytes(bytes([randrange(256) for _ in range(2**18)]))
+            file.chmod(0o644)
 
+        subprocess.run(['ostree', 'init', '--mode=archive-z2',
+                        '--repo={}'.format(path)], check=True)
+        subprocess.run(['ostree', '--repo={}'.format(path), 'commit',
+                        '--consume', '--branch=master',
+                        '--owner-uid=0', '--owner-gid=0', '--no-xattrs',
+                        '--timestamp=1970-01-01 00:00:00 +0000', str(tree)],
+                       check=True)
+    except PermissionError:
+        time.sleep(100)
+
+
+parser = argparse.ArgumentParser()
+parser.add_argument('-p', '--port', type=int, required=True,
+                    help='listening port')
+parser.add_argument('-c', '--create', action='store_true',
+                    help='create new ostree repo')
+parser.add_argument('-d', '--dir', help='ostree repo directory')
+parser.add_argument('-f', '--fail', type=int, help='fail every nth request')
+parser.add_argument('-s', '--sleep', type=float,
+                    help='sleep for n.n seconds for every GET request')
+parser.add_argument('-t', '--tls', action='store_true',
+                    help='require TLS from clients')
+args = parser.parse_args()
+
+signal.signal(signal.SIGTERM, sig_handler)
 try:
-    httpd.serve_forever()
-except KeyboardInterrupt as k:
+    with ExitStack() as stack:
```

Comment:
What benefit does the ExitStack and sig_handler provide? Just curious, not doubting. :)

Reply:
sys.exit(0) in signal handler will raise a SystemExit exception, which will unwind the ExitStack cleaning up the tmp directory